### PR TITLE
Add documentation / examples for ServiceProduct usage

### PR DESCRIPTION
**Pull request by kubukoz**
Includes changes from #1053 by @miguel-vila.

image

### CODE REVIEW

File: modules/bootstrapped/src/generated/smithy4s/example/product/ExampleOperationInput.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package smithy4s.example.product
```
Comment (kubukoz) [inline modules/bootstrapped/src/generated/smithy4s/example/product/ExampleOperationInput.scala:1]: **thought:** side note, I think the recent codegen changes around the `bootstrapped` module could've broken something wrt. detecting stale files - e.g. if you delete a shape from the smithy specs, you need to run a `clean` in order for that to be reflected on the filesystem

File: modules/bootstrapped/src/generated/smithy4s/example/product/ObjectService.scala
Comment (kubukoz) [inline modules/bootstrapped/src/generated/smithy4s/example/product/ObjectService.scala:1]: **thought:** removed this one in favor of the `ExampleService` that @miguel-vila added in the docs in #1053

File: modules/docs/markdown/04-codegen/01-customisation/10-service-product.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+      override def apply[I, E, O, SI, SO](
+        fa: sp.service.Endpoint[I, E, O, SI, SO]
+      ): List[I] => List[O] = _.map(in => interp(fa.wrap(in)))
+
+    },
+  )
+
+// Usage
+
+listClient(impl)( /* implicit scope problem here - TODO */ ExampleService.serviceProduct)
```
Comment (kubukoz) [inline modules/docs/markdown/04-codegen/01-customisation/10-service-product.md:134]: **issue:** we're lacking something that would make this visible implicitly.

I tried the following (all combinations):

- making `serviceProduct` an implicit val in the interface
- making `serviceProduct` an implicit val in the generated code

I'm not sure why it's not being detected.

File: modules/docs/markdown/04-codegen/01-customisation/10-service-product.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+              }
+
+          }
+
+      },
+    )
+
+  def build: Alg[ToList] = sp
+    .service
+    .algebra(new sp.service.EndpointCompiler[ToList] {
```
Comment (kubukoz) [inline modules/docs/markdown/04-codegen/01-customisation/10-service-product.md:191]: **praise:** for @Baccata - the whole `algebra` / `EndpointCompiler` thing is very cool.

File: modules/bootstrapped/src/generated/smithy4s/example/product/ExampleOperationInput.scala (hunk 1 above)
Comment (Baccata) [reply modules/bootstrapped/src/generated/smithy4s/example/product/ExampleOperationInput.scala:1]: >  e.g. if you delete a shape from the smithy specs, you need to run a clean in order for that to be reflected on the filesystem 

That is correct. But at least now it's wired in `clean`, previously you had to manually remove stuff.

File: modules/docs/markdown/04-codegen/01-customisation/10-service-product.md (hunk 2 above)
Comment (Baccata) [reply modules/docs/markdown/04-codegen/01-customisation/10-service-product.md:134]: > making serviceProduct an implicit val in the generated code 

where was the implicit located when you tried ?

Comment (kubukoz) [reply modules/docs/markdown/04-codegen/01-customisation/10-service-product.md:134]: exactly where the non-implicit val is (companion object of `ExampleServiceGen`)

**Comment by Baccata** [comment]
I'm gonna merge it as is, it's good and we can amend later if we find a way to make the implicit work.

**Review by Baccata — APPROVED** [review]

**Pull request merged**